tests/main/disk-space-awareness/task.yaml

```diff
@@ -0,0 +1,72 @@
+summary: Check that basic snap operation are aware of low disk space.
+
+# this test is remounts /var/lib/snapd which is too intrusive for core and
+# selinux, disabled on these systems.
+systems: [-fedora-*, -amazon-*, -centos-*, -ubuntu-core-*]
+
+description: |
+  Check that operations such as snap installation or snap removal (with an
+  automatic snapshot) error out early if there is not enough disk space.
+
+environment:
+  TMPFSMOUNT: /var/lib/snapd
+
+prepare: |
+  systemctl stop snapd.{socket,service}
+
+  # mount /var/lib/snapd on a tmpfs
+  mount -t tmpfs tmpfs -o size=500M,mode=0755 "$TMPFSMOUNT"
+
+  systemctl start snapd.{socket,service}
+
+restore: |
+  systemctl stop snapd.{socket,service}
+  umount -l "$TMPFSMOUNT"
+  systemctl start snapd.{socket,service}
+
+execute: |
+  resize() {
+    # remount can increase/decrease the size as long as it is enough for
+    # already stored data.
+    mount -o remount,size="$1",mode=0755 "$TMPFSMOUNT"
+  }
+
+  # disk usage, in MB
+  get_disk_usage() {
+    df -B1048576 "$TMPFSMOUNT" | tail -1 | awk '{print $3}'
+  }
+
+  echo "Enabling experimental disk space checks"
+  snap set system experimental.check-disk-space-install=true
+  snap set system experimental.check-disk-space-remove=true
+
+  echo "Snap download fails with little disk space"
+  NEWSIZE=$(($(get_disk_usage) + 1))
+  resize "${NEWSIZE}M"
+  snap install hello-world 2>&1 | MATCH  "cannot install \"hello-world\" due to low disk space"
+
+  # note, installing hello-world also pulls core snap.
+  echo "And succeeds with enough space"
+  resize 600M
```

Comment:
This may fail on smaller systems, is that really the minimum that works?

Reply:
No really, ~100M is probably a minimum (core snap + test snaps), I just wanted to be safe. I can reduce it in a followup.